middleware.go

```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
+
+// RetryMiddleware allows retrying of certain retryable http errors.
+// This only handles very specific status codes, ones that are deemed retryable;
+//
+//  - 502 Bad Gateway
+//  - 503 Service Unavailable
+//  - 504 Gateway Timeout
+type retryMiddleware struct {
+	policy              RetryPolicy
+	wrappedRoundTripper http.RoundTripper
+	clock               clock.Clock
+}
+
+type RetryPolicy struct {
+	Delay    time.Duration
+	Attempts int
+}
+
+// Validate validates the RetryPolicy for any issues.
+func (p RetryPolicy) Validate() error {
+	if p.Attempts < 1 {
+		return errors.Errorf("expected at least one attempt")
+	}
+	return nil
+}
+
+// makeRetryMiddleware creates a retry transport.
+func makeRetryMiddleware(transport http.RoundTripper, policy RetryPolicy, clock clock.Clock) http.RoundTripper {
+	return retryMiddleware{
+		policy:              policy,
+		wrappedRoundTripper: transport,
+		clock:               clock,
+	}
+}
+
+type retryableErr struct{}
+
+func (retryableErr) Error() string {
+	return "retryable error"
+}
+
+func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	var res *http.Response
+	err := retry.Call(retry.CallArgs{
+		Clock: m.clock,
+		Func: func() error {
+			if err := req.Context().Err(); err != nil {
+				return err
+			}
+
+			var retryable bool
+			var err error
+			res, retryable, err = m.roundTrip(req)
+			if err != nil {
+				return err
+			}
+			if retryable {
+				return retryableErr{}
+			}
+			return nil
+		},
+		IsFatalError: func(err error) bool {
+			// Work out if it's not a retryable error.
+			_, ok := errors.Cause(err).(retryableErr)
+			return !ok
+		},
+		Attempts: m.policy.Attempts,
+		Delay:    m.policy.Delay,
+		BackoffFunc: func(delay time.Duration, attempts int) time.Duration {
+			return m.defaultBackoff(res, delay)
+		},
+	})
+
+	return res, err
+}
+
+func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, error) {
+	res, err := m.wrappedRoundTripper.RoundTrip(req)
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch res.StatusCode {
+	case http.StatusBadGateway, http.StatusGatewayTimeout:
+		// The request should be retryable.
+		fallthrough
+	case http.StatusServiceUnavailable, http.StatusTooManyRequests:
+		// The request should be retryable, but additionally should contain
+		// a potential Retry-After header.
+		return res, true, nil
+	default:
+		// Don't handle any of the following status codes.
+		return res, false, nil
+	}
+}
+
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
+func (m retryMiddleware) defaultBackoff(resp *http.Response, max time.Duration) time.Duration {
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusTooManyRequests {
+		if header := resp.Header.Get("Retry-After"); header != "" {
+			// Attempt to parse the header from the request.
+			retry, err := strconv.ParseInt(header, 10, 64)
```

Comment:
Found the hard way, the retry value can also be in time.RFC1123 format: https://github.com/go-goose/goose/blob/v2/http/client.go#L384-L409

Reply:
NICE, thanks for this. I wonder if we can maybe drop this one in goose eventually as we can pass in a juju/http client? Either way this is great information.

Fixed in e47bd4b

Reply:
https://github.com/juju/http/pull/6/files#diff-02f50901140d6057da6310a106670552aa766a093efbc2200fb34c099b762131R200-R202